### PR TITLE
Report nested nullability warnings for arguments to user-defined conversions

### PR DESCRIPTION
**Pull request by cston**
Fixes https://github.com/dotnet/roslyn/issues/31770.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 1]
```diff
@@ -3315,6 +3318,7 @@ private bool HasTopLevelNullabilityConversion(TypeSymbolWithAnnotations source,
             Debug.Assert(node != null);
             Debug.Assert(operandOpt != null || !operandType.IsNull);
             Debug.Assert(!targetTypeWithNullability.IsNull);
+            Debug.Assert((object)target != null || assignmentKind != AssignmentKind.Argument);
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: Is there really no way for an AssignmentKind.Return to get here with no Target? While we don't have any tests that are triggering this, it seems like a return statement with a conversion should get here.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 2]
```diff
@@ -568,20 +570,20 @@ private bool ReportNullableAssignmentIfNecessary(BoundExpression value, TypeSymb
 
             // Report warning converting null literal to non-nullable reference type.
             // target (e.g.: `object x = null;` or calling `void F(object y)` with `F(null)`).
-            bool reportNullLiteralAssignmentIfNecessary()
+            bool reportNullLiteralAssignmentIfNecessary(BoundExpression expr)
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: Other than making this a candidate for a `static` local function when we get them, did this actually add any value? Not criticizing, just making sure I'm not missing something.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 1 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: Sorry, scratch this, misread the assert.

---
In reply to: [241615041](https://github.com/dotnet/roslyn/pull/31771#discussion_r241615041) [](ancestors = 241615041)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 3]
```diff
@@ -65178,5 +65178,35 @@ static void F3(int? z, int? w)
...
+    static void Main()
+    {
+        A<object?> a = new A<object?>();
+        B b = a; // 1
+        F(a); // 2
+    }
+}";
+            var comp = CreateCompilation(source, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics(
+                // (12,15): warning CS8619: Nullability of reference types in value of type 'A<object?>' doesn't match target type 'A<object>'.
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: Is there any way we can make these errors more informative? It's relatively obvious in this simple example, but there's no mention in this error about the difference being part of the conversion, which in a more complicated scenario would be in another section of the code written by a different person.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 2 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: Passing the expression as an argument allows null analysis to verify that `value` is not null in this method (see [https://github.com/dotnet/roslyn/pull/31719](https://github.com/dotnet/roslyn/pull/31719)).

---
In reply to: [241615378](https://github.com/dotnet/roslyn/pull/31771#discussion_r241615378) [](ancestors = 241615378)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 3 above)
Comment (cston) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: Logged https://github.com/dotnet/roslyn/issues/31798.

Comment (cston) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: Logged [https://github.com/dotnet/roslyn/issues/31798](https://github.com/dotnet/roslyn/issues/31798).

---
In reply to: [241617155](https://github.com/dotnet/roslyn/pull/31771#discussion_r241617155) [](ancestors = 241617155)

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 4]
```diff
@@ -3040,6 +3042,7 @@ private static Symbol AsMemberOfResultType(TypeSymbolWithAnnotations resultType,
 
         private static Symbol AsMemberOfType(NamedTypeSymbol containingType, Symbol symbol)
         {
+            Debug.Assert((object)symbol != null);
             if (symbol is null)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >if (symbol is null) [](start = 12, length = 19)

Would it make sense to remove this null check then?

**Pull request merged**